File: sdks/node/src/plugins/qwen-sources.ts
import os from 'node:os';
import path from 'node:path';
import fsp from 'node:fs/promises';
import { parse as parseDotenv } from 'dotenv';
import {
  createQwenDocument,
  parseQwenDocument,
  parseQwenJsoncObject,
  qwenDocumentLabel,
  qwenSourceLabel,
  type QwenDocument,
  type QwenDocumentKind,
} from './qwen-config.js';
import { readPhysicalFile, type FileSnapshot } from './managed-files.js';

const MAX_SOURCE_BYTES = 2 * 1024 * 1024;
const HOME_ENV_KEYS = ['QWEN_HOME', 'QWEN_RUNTIME_DIR'] as const;

export interface QwenSourcePrecondition {
  readonly filePath: string;
  readonly label: string;
  readonly maximumBytes: number;
  readonly original?: FileSnapshot;
}

export interface QwenDisableControl {
  readonly disabled: boolean;
  readonly source?: QwenDocument;
}

export interface QwenSources {
  readonly qwenHome: string;
  readonly systemDefaults: QwenDocument;
  readonly user: QwenDocument;
  readonly workspace: QwenDocument;
  readonly system: QwenDocument;
  readonly workspaceActive: boolean;
  readonly workspaceTrusted: boolean;
  readonly disableControl: QwenDisableControl;
  readonly preconditions: readonly QwenSourcePrecondition[];
}

interface RoutingResult {
  readonly qwenHome: string;
  readonly preconditions: readonly QwenSourcePrecondition[];
}

function defaultQwenHome(): string {
  const home = os.homedir();
  return home ? path.join(home, '.qwen') : path.join(os.tmpdir(), '.qwen');
}

function resolveConfigPath(value: string): string {
  let resolved = value;
  if (resolved === '~' || resolved.startsWith('~/') || resolved.startsWith('~\\')) {
    const segments = resolved === '~'
      ? []
      : resolved.slice(2).split(/[/\\]+/).filter(Boolean);
    resolved = path.join(os.homedir(), ...segments);
  }
  return path.isAbsolute(resolved) ? resolved : path.resolve(resolved);
}

function environmentValue(key: string): { own: boolean; value?: string } {
  return {
    own: Object.hasOwn(process.env, key),
    value: process.env[key],
  };
}

async function resolveQwenRouting(): Promise<RoutingResult> {
  const values = new Map<string, string | undefined>();
  const owned = new Set<string>();
  for (const key of HOME_ENV_KEYS) {
    const current = environmentValue(key);
    values.set(key, current.value);
    if (current.own) owned.add(key);
  }
  if (HOME_ENV_KEYS.every((key) => values.get(key))) {
    return {
      qwenHome: resolveConfigPath(values.get('QWEN_HOME')!),
      preconditions: [],
    };
  }

  const initialQwenHome = values.get('QWEN_HOME');
  const initialDirectory = initialQwenHome
    ? resolveConfigPath(initialQwenHome)
    : defaultQwenHome();
  const candidates = [path.join(initialDirectory, '.env')];
  if (!initialQwenHome) candidates.push(path.join(path.dirname(initialDirectory), '.env'));
  const preconditions: QwenSourcePrecondition[] = [];
  const visited = new Set<string>();

  const readCandidate = async (filePath: string): Promise<void> => {
    const resolved = path.resolve(filePath);
    const key = process.platform === 'win32' ? resolved.toLowerCase() : resolved;
    if (visited.has(key)) return;
    visited.add(key);
    const snapshot = await readPhysicalFile(resolved, 'Qwen Code home environment');
    preconditions.push({
      filePath: resolved,
      label: 'Qwen Code home environment',
      maximumBytes: MAX_SOURCE_BYTES,
      original: snapshot,
    });
    if (!snapshot) return;
    const parsed = parseDotenv(snapshot.contents);
    for (const envKey of HOME_ENV_KEYS) {
      const value = parsed[envKey];
      if (value && !owned.has(envKey)) {
        values.set(envKey, value);
        owned.add(envKey);
      }
    }
  };

  for (const candidate of candidates) await readCandidate(candidate);
  const discoveredHome = values.get('QWEN_HOME');
  if (discoveredHome && discoveredHome !== initialQwenHome) {
    const discoveredDirectory = resolveConfigPath(discoveredHome);
    if (discoveredDirectory !== initialDirectory) {
      await readCandidate(path.join(discoveredDirectory, '.env'));
    }
  }
  return {
    qwenHome: values.get('QWEN_HOME')
      ? resolveConfigPath(values.get('QWEN_HOME')!)
      : defaultQwenHome(),
    preconditions,
  };
}

function systemSettingsPath(): string {
  const configured = process.env.QWEN_CODE_SYSTEM_SETTINGS_PATH;
  if (configured) return path.resolve(configured);
  if (process.platform === 'darwin') {
    return '/Library/Application Support/QwenCode/settings.json';
  }
  if (process.platform === 'win32') return 'C:\\ProgramData\\qwen-code\\settings.json';
  return '/etc/qwen-code/settings.json';
}

function systemDefaultsPath(systemPath: string): string {
  const configured = process.env.QWEN_CODE_SYSTEM_DEFAULTS_PATH;
  return configured
    ? path.resolve(configured)
    : path.join(path.dirname(systemPath), 'system-defaults.json');
}

async function readDocument(
  kind: QwenDocumentKind,
  filePath: string,
): Promise<QwenDocument> {
  const snapshot = await readPhysicalFile(filePath, qwenSourceLabel(kind));
  return snapshot
    ? parseQwenDocument({
      kind,
      exists: true,
      filePath,
      raw: snapshot.contents,
      snapshot,
    })
    : createQwenDocument(kind, filePath);
}

async function canonicalPath(filePath: string): Promise<string> {
  try {
    return await fsp.realpath(filePath);
  } catch {
    return path.resolve(filePath);
  }
}

function comparisonPath(filePath: string): string {
  const resolved = path.resolve(filePath);
  return process.platform === 'win32' ? resolved.toLowerCase() : resolved;
}

function isWithin(child: string, parent: string): boolean {
  const relative = path.relative(comparisonPath(parent), comparisonPath(child));
  return relative === '' || (!relative.startsWith('..') && !path.isAbsolute(relative));
}

async function workspaceTrust(
  system: QwenDocument,
  user: QwenDocument,
  qwenHome: string,
  workspacePath: string,
): Promise<{ trusted: boolean; precondition?: QwenSourcePrecondition }> {
  const enabled = user.folderTrustEnabled ?? system.folderTrustEnabled ?? false;
  if (!enabled) return { trusted: true };
  const configured = process.env.QWEN_CODE_TRUSTED_FOLDERS_PATH;
  const filePath = configured
    ? path.resolve(configured)
    : path.join(qwenHome, 'trustedFolders.json');
  const snapshot = await readPhysicalFile(filePath, 'Qwen Code trusted folders');
  const precondition = {
    filePath,
    label: 'Qwen Code trusted folders',
    maximumBytes: MAX_SOURCE_BYTES,
    original: snapshot,
  };
  if (!snapshot) return { trusted: true, precondition };
  const rules = parseQwenJsoncObject(snapshot.contents, `Qwen Code trusted folders at ${filePath}`);
  for (const [rulePath, level] of Object.entries(rules)) {
    if (!['TRUST_FOLDER', 'TRUST_PARENT', 'DO_NOT_TRUST'].includes(String(level))) {
      throw new Error(`Qwen Code trusted folders has invalid trust level for "${rulePath}"`);
    }
  }
  const workspace = await canonicalPath(workspacePath);
  for (const [rulePath, level] of Object.entries(rules)) {
    const canonicalRule = await canonicalPath(String(rulePath));
    const trustRoot = level === 'TRUST_PARENT' ? path.dirname(canonicalRule) : canonicalRule;
    if ((level === 'TRUST_FOLDER' || level === 'TRUST_PARENT')
      && isWithin(workspace, trustRoot)) return { trusted: true, precondition };
  }
  for (const [rulePath, level] of Object.entries(rules)) {
    if (level === 'DO_NOT_TRUST'
      && comparisonPath(workspace) === comparisonPath(await canonicalPath(rulePath))) {
      return { trusted: false, precondition };
    }
  }
  return { trusted: true, precondition };
}

function effectiveDisable(
  systemDefaults: QwenDocument,
  user: QwenDocument,
  workspace: QwenDocument,
  system: QwenDocument,
  useWorkspace: boolean,
): QwenDisableControl {
  let disabled = false;
  let source: QwenDocument | undefined;
  const documents = [systemDefaults, user, ...(useWorkspace ? [workspace] : []), system];
  for (const document of documents) {
    if (document.disableAllHooks === undefined) continue;
    disabled = document.disableAllHooks;
    source = document;
  }
  return { disabled, source };
}

function sourcePrecondition(document: QwenDocument): QwenSourcePrecondition {
  return {
    filePath: document.filePath,
    label: qwenDocumentLabel(document),
    maximumBytes: MAX_SOURCE_BYTES,
    original: document.snapshot,
  };
}

function deduplicatePreconditions(
  values: readonly QwenSourcePrecondition[],
): QwenSourcePrecondition[] {
  const result = new Map<string, QwenSourcePrecondition>();
  for (const value of values) {
    const key = comparisonPath(value.filePath);
    if (!result.has(key)) result.set(key, value);
  }
  return [...result.values()];
}

export async function readQwenSources(): Promise<QwenSources> {
  const routing = await resolveQwenRouting();
  const systemPath = systemSettingsPath();
  const workspacePath = path.join(process.cwd(), '.qwen', 'settings.json');
  const system = await readDocument('system', systemPath);
  const systemDefaults = await readDocument(
    'system-defaults',
    systemDefaultsPath(systemPath),
  );
  const user = await readDocument('user', path.join(routing.qwenHome, 'settings.json'));
  const [canonicalWorkspace, canonicalHome] = await Promise.all([
    canonicalPath(process.cwd()),
    canonicalPath(os.homedir()),
  ]);
  const workspaceActive = comparisonPath(canonicalWorkspace) !== comparisonPath(canonicalHome);
  const workspace = workspaceActive
    ? await readDocument('workspace', workspacePath)
    : createQwenDocument('workspace', workspacePath);
  const trust = workspaceActive
    ? await workspaceTrust(system, user, routing.qwenHome, canonicalWorkspace)
    : { trusted: false as const };
  const useWorkspace = workspaceActive && trust.trusted;
  const preconditions = deduplicatePreconditions([
    ...routing.preconditions,
    sourcePrecondition(systemDefaults),
    sourcePrecondition(user),
    ...(workspaceActive ? [sourcePrecondition(workspace)] : []),
    sourcePrecondition(system),
    ...(trust.precondition ? [trust.precondition] : []),
  ]);
  return {
    qwenHome: routing.qwenHome,
    systemDefaults,
    user,
    workspace,
    system,
    workspaceActive,
    workspaceTrusted: trust.trusted,
    disableControl: effectiveDisable(
      systemDefaults,
      user,
      workspace,
      system,
      useWorkspace,
    ),
    preconditions,
  };
}

export function requireQwenHooksEnabled(sources: QwenSources): void {
  if (!sources.disableControl.disabled) return;
  const source = sources.disableControl.source;
  const location = source ? `${qwenDocumentLabel(source)} at ${source.filePath}` : 'effective settings';
  throw new Error(`Qwen Code hooks are disabled by disableAllHooks in ${location}`);
}

File: sdks/go/cmd/elydora/plugins/qwen_sources.go
package plugins

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

var qwenHomeEnvironmentKeys = [...]string{"QWEN_HOME", "QWEN_RUNTIME_DIR"}

type qwenDisableControl struct {
	disabled bool
	source   *qwenDocument
}

type qwenSources struct {
	qwenHome         string
	systemDefaults   *qwenDocument
	user             *qwenDocument
	workspace        *qwenDocument
	system           *qwenDocument
	workspaceActive  bool
	workspaceTrusted bool
	disableControl   qwenDisableControl
	preconditions    []filePrecondition
}

type qwenRoutingResult struct {
	qwenHome      string
	preconditions []filePrecondition
}

func defaultQwenHome() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", fmt.Errorf("resolve home directory: %w", err)
	}
	return filepath.Join(home, ".qwen"), nil
}

func resolveQwenStoragePath(value string) (string, error) {
	resolved := value
	if value == "~" || len(value) >= 2 && value[0] == '~' &&
		(value[1] == '/' || value[1] == '\\') {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", fmt.Errorf("resolve home directory: %w", err)
		}
		resolved = home
		if value != "~" {
			for _, segment := range qwenPathSeparatorPattern.Split(value[2:], -1) {
				if segment != "" {
					resolved = filepath.Join(resolved, segment)
				}
			}
		}
	}
	if filepath.IsAbs(resolved) {
		return filepath.Clean(resolved), nil
	}
	absolute, err := filepath.Abs(resolved)
	if err != nil {
		return "", fmt.Errorf("resolve Qwen Code storage path %q: %w", value, err)
	}
	return absolute, nil
}

func qwenComparisonPath(filePath string) string {
	resolved, err := filepath.Abs(filePath)
	if err != nil {
		resolved = filepath.Clean(filePath)
	}
	if runtime.GOOS == "windows" {
		return strings.ToLower(resolved)
	}
	return resolved
}

func resolveQwenRouting() (*qwenRoutingResult, error) {
	values := map[string]string{}
	owned := map[string]bool{}
	for _, key := range qwenHomeEnvironmentKeys {
		value, present := os.LookupEnv(key)
		values[key] = value
		owned[key] = present
	}
	if values["QWEN_HOME"] != "" && values["QWEN_RUNTIME_DIR"] != "" {
		home, err := resolveQwenStoragePath(values["QWEN_HOME"])
		return &qwenRoutingResult{qwenHome: home}, err
	}
	defaultHome, err := defaultQwenHome()
	if err != nil {
		return nil, err
	}
	initialHome := values["QWEN_HOME"]
	initialDirectory := defaultHome
	if initialHome != "" {
		initialDirectory, err = resolveQwenStoragePath(initialHome)
		if err != nil {
			return nil, err
		}
	}
	candidates := []string{filepath.Join(initialDirectory, ".env")}
	if initialHome == "" {
		candidates = append(candidates, filepath.Join(filepath.Dir(initialDirectory), ".env"))
	}
	preconditions := make([]filePrecondition, 0, 3)
	visited := map[string]bool{}
	readCandidate := func(filePath string) error {
		resolved, resolveErr := filepath.Abs(filePath)
		if resolveErr != nil {
			return fmt.Errorf("resolve Qwen Code home environment path: %w", resolveErr)
		}
		key := qwenComparisonPath(resolved)
		if visited[key] {
			return nil
		}
		visited[key] = true
		snapshot, readErr := readManagedFile(
			resolved,
			"Qwen Code home environment",
			maxManagedSourceBytes,
		)
		if readErr != nil {
			return readErr
		}
		preconditions = append(preconditions, filePrecondition{
			filePath: resolved, label: "Qwen Code home environment",
			snapshot: snapshot, maximumSize: maxManagedSourceBytes,
		})
		if snapshot == nil {
			return nil
		}
		parsed := parseDotenv(snapshot.contents)
		for _, envKey := range qwenHomeEnvironmentKeys {
			value := parsed[envKey]
			if value != "" && !owned[envKey] {
				values[envKey] = value
				owned[envKey] = true
			}
		}
		return nil
	}
	for _, candidate := range candidates {
		if err := readCandidate(candidate); err != nil {
			return nil, err
		}
	}
	discoveredHome := values["QWEN_HOME"]
	if discoveredHome != "" && discoveredHome != initialHome {
		discoveredDirectory, resolveErr := resolveQwenStoragePath(discoveredHome)
		if resolveErr != nil {
			return nil, resolveErr
		}
		if !sameQwenPath(discoveredDirectory, initialDirectory) {
			if err := readCandidate(filepath.Join(discoveredDirectory, ".env")); err != nil {
				return nil, err
			}
		}
	}
	resolvedHome := defaultHome
	if values["QWEN_HOME"] != "" {
		resolvedHome, err = resolveQwenStoragePath(values["QWEN_HOME"])
		if err != nil {
			return nil, err
		}
	}
	return &qwenRoutingResult{
		qwenHome:      resolvedHome,
		preconditions: preconditions,
	}, nil
}

func resolveQwenHome() (string, error) {
	routing, err := resolveQwenRouting()
	if err != nil {
		return "", err
	}
	return routing.qwenHome, nil
}

func qwenSystemSettingsPath() (string, error) {
	if configured := os.Getenv("QWEN_CODE_SYSTEM_SETTINGS_PATH"); configured != "" {
		return filepath.Abs(configured)
	}
	switch runtime.GOOS {
	case "darwin":
		return "/Library/Application Support/QwenCode/settings.json", nil
	case "windows":
		return `C:\ProgramData\qwen-code\settings.json`, nil
	default:
		return "/etc/qwen-code/settings.json", nil
	}
}

func qwenSystemDefaultsPath(systemPath string) (string, error) {
	if configured := os.Getenv("QWEN_CODE_SYSTEM_DEFAULTS_PATH"); configured != "" {
		return filepath.Abs(configured)
	}
	return filepath.Join(filepath.Dir(systemPath), "system-defaults.json"), nil
}

func readQwenDocument(kind, filePath string) (*qwenDocument, error) {
	snapshot, err := readManagedFile(filePath, qwenSourceLabel(kind), maxManagedSourceBytes)
	if err != nil {
		return nil, err
	}
	if snapshot == nil {
		return createQwenDocument(kind, filePath)
	}
	return parseQwenDocument(kind, true, filePath, snapshot.contents, snapshot)
}

func canonicalQwenPath(filePath string) string {
	canonical, err := filepath.EvalSymlinks(filePath)
	if err == nil {
		return canonical
	}
	absolute, absErr := filepath.Abs(filePath)
	if absErr == nil {
		return absolute
	}
	return filepath.Clean(filePath)
}

func qwenPathWithin(child, parent string) bool {
	relative, err := filepath.Rel(qwenComparisonPath(parent), qwenComparisonPath(child))
	if err != nil {
		return false
	}
	return relative == "." ||
		(relative != ".." && !strings.HasPrefix(relative, ".."+string(filepath.Separator)) &&
			!filepath.IsAbs(relative))
}

func qwenWorkspaceTrust(
	system, user *qwenDocument,
	qwenHome, workspacePath string,
) (bool, *filePrecondition, error) {
	var enabled *bool
	if user != nil && user.folderTrustEnabled != nil {
		enabled = user.folderTrustEnabled
	} else if system != nil {
		enabled = system.folderTrustEnabled
	}
	if enabled == nil || !*enabled {
		return true, nil, nil
	}
	filePath := os.Getenv("QWEN_CODE_TRUSTED_FOLDERS_PATH")
	var err error
	if filePath != "" {
		filePath, err = filepath.Abs(filePath)
		if err != nil {
			return false, nil, fmt.Errorf("resolve Qwen Code trusted folders path: %w", err)
		}
	} else {
		filePath = filepath.Join(qwenHome, "trustedFolders.json")
	}
	snapshot, err := readManagedFile(
		filePath,
		"Qwen Code trusted folders",
		maxManagedSourceBytes,
	)
	if err != nil {
		return false, nil, err
	}
	precondition := &filePrecondition{
		filePath: filePath, label: "Qwen Code trusted folders",
		snapshot: snapshot, maximumSize: maxManagedSourceBytes,
	}
	if snapshot == nil {
		return true, precondition, nil
	}
	rules, err := decodeJSONCObject(
		snapshot.contents,
		fmt.Sprintf("Qwen Code trusted folders at %s", filePath),
		false,
	)
	if err != nil {
		return false, nil, err
	}
	for rulePath, level := range rules {
		if level != "TRUST_FOLDER" && level != "TRUST_PARENT" && level != "DO_NOT_TRUST" {
			return false, nil, fmt.Errorf(
				"Qwen Code trusted folders has invalid trust level for %q",
				rulePath,
			)
		}
	}
	workspace := canonicalQwenPath(workspacePath)
	for rulePath, level := range rules {
		canonicalRule := canonicalQwenPath(rulePath)
		trustRoot := canonicalRule
		if level == "TRUST_PARENT" {
			trustRoot = filepath.Dir(canonicalRule)
		}
		if (level == "TRUST_FOLDER" || level == "TRUST_PARENT") &&
			qwenPathWithin(workspace, trustRoot) {
			return true, precondition, nil
		}
	}
	for rulePath, level := range rules {
		if level == "DO_NOT_TRUST" &&
			qwenComparisonPath(workspace) == qwenComparisonPath(canonicalQwenPath(rulePath)) {
			return false, precondition, nil
		}
	}
	return true, precondition, nil
}

func effectiveQwenDisable(
	systemDefaults, user, workspace, system *qwenDocument,
	useWorkspace bool,
) qwenDisableControl {
	control := qwenDisableControl{}
	documents := []*qwenDocument{systemDefaults, user}
	if useWorkspace {
		documents = append(documents, workspace)
	}
	documents = append(documents, system)
	for _, document := range documents {
		if document == nil || document.disableAllHooks == nil {
			continue
		}
		control.disabled = *document.disableAllHooks
		control.source = document
	}
	return control
}

func qwenSourcePrecondition(document *qwenDocument) filePrecondition {
	return filePrecondition{
		filePath:    document.filePath,
		label:       qwenDocumentLabel(document),
		snapshot:    document.snapshot,
		maximumSize: maxManagedSourceBytes,
	}
}

func deduplicateQwenPreconditions(values []filePrecondition) []filePrecondition {
	result := make([]filePrecondition, 0, len(values))
	seen := map[string]bool{}
	for _, value := range values {
		key := qwenComparisonPath(value.filePath)
		if seen[key] {
			continue
		}
		seen[key] = true
		result = append(result, value)
	}
	return result
}

func readQwenSources() (*qwenSources, error) {
	routing, err := resolveQwenRouting()
	if err != nil {
		return nil, err
	}
	systemPath, err := qwenSystemSettingsPath()
	if err != nil {
		return nil, err
	}
	defaultsPath, err := qwenSystemDefaultsPath(systemPath)
	if err != nil {
		return nil, err
	}
	workspaceDirectory, err := os.Getwd()
	if err != nil {
		return nil, fmt.Errorf("resolve Qwen Code workspace: %w", err)
	}
	workspacePath := filepath.Join(workspaceDirectory, ".qwen", "settings.json")
	system, err := readQwenDocument(qwenSystemKind, systemPath)
	if err != nil {
		return nil, err
	}
	systemDefaults, err := readQwenDocument(qwenSystemDefaultsKind, defaultsPath)
	if err != nil {
		return nil, err
	}
	user, err := readQwenDocument(
		qwenUserKind,
		filepath.Join(routing.qwenHome, "settings.json"),
	)
	if err != nil {
		return nil, err
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return nil, fmt.Errorf("resolve home directory: %w", err)
	}
	canonicalWorkspace := canonicalQwenPath(workspaceDirectory)
	workspaceActive := qwenComparisonPath(canonicalWorkspace) !=
		qwenComparisonPath(canonicalQwenPath(home))
	workspace, err := createQwenDocument(qwenWorkspaceKind, workspacePath)
	if err != nil {
		return nil, err
	}
	if workspaceActive {
		workspace, err = readQwenDocument(qwenWorkspaceKind, workspacePath)
		if err != nil {
			return nil, err
		}
	}
	workspaceTrusted := false
	var trustPrecondition *filePrecondition
	if workspaceActive {
		workspaceTrusted, trustPrecondition, err = qwenWorkspaceTrust(
			system,
			user,
			routing.qwenHome,
			canonicalWorkspace,
		)
		if err != nil {
			return nil, err
		}
	}
	if err := validateQwenJavaScriptMatchers([]qwenHookSettings{
		systemDefaults.hooks,
		user.hooks,
		workspace.hooks,
		system.hooks,
	}); err != nil {
		return nil, err
	}
	preconditions := append([]filePrecondition(nil), routing.preconditions...)
	preconditions = append(
		preconditions,
		qwenSourcePrecondition(systemDefaults),
		qwenSourcePrecondition(user),
	)
	if workspaceActive {
		preconditions = append(preconditions, qwenSourcePrecondition(workspace))
	}
	preconditions = append(preconditions, qwenSourcePrecondition(system))
	if trustPrecondition != nil {
		preconditions = append(preconditions, *trustPrecondition)
	}
	return &qwenSources{
		qwenHome:         routing.qwenHome,
		systemDefaults:   systemDefaults,
		user:             user,
		workspace:        workspace,
		system:           system,
		workspaceActive:  workspaceActive,
		workspaceTrusted: workspaceTrusted,
		disableControl: effectiveQwenDisable(
			systemDefaults,
			user,
			workspace,
			system,
			workspaceActive && workspaceTrusted,
		),
		preconditions: deduplicateQwenPreconditions(preconditions),
	}, nil
}

func requireQwenHooksEnabled(sources *qwenSources) error {
	if sources == nil {
		return fmt.Errorf("Qwen Code settings sources are required")
	}
	if !sources.disableControl.disabled {
		return nil
	}
	location := "effective settings"
	if sources.disableControl.source != nil {
		location = fmt.Sprintf(
			"%s at %s",
			qwenDocumentLabel(sources.disableControl.source),
			sources.disableControl.source.filePath,
		)
	}
	return fmt.Errorf(
		"Qwen Code hooks are disabled by disableAllHooks in %s",
		location,
	)
}
